Branch pybind11 can not be compiled without GPU.

FYI.

`g++ -c -std=c++11 -I.. -isystem /home/work/kaldi/tools/openfst-1.6.7/include -O1 -Wno-sign-compare -Wall -Wno-sign-compare -Wno-unused-local-typedefs -Wno-deprecated-declarations -Winit-self -DKALDI_DOUBLEPRECISION=0 -DHAVE_EXECINFO_H=1 -DHAVE_CXXABI_H -DHAVE_MKL -I/opt/intel/mkl/include -m64 -msse -msse2 -pthread -g -fPIC -I/usr/include/python3.6m -I/usr/local/include/python3.6m -I/root/.local/include/python3.6m -O3 -flto -I. -o dlpack/dlpack_pybind.o dlpack/dlpack_pybind.cc
dlpack/dlpack_pybind.cc: In function ‘kaldi::DLPackCuSubVector<float>* kaldi::CuSubVectorFromDLPack(pybind11::capsule*)’:
dlpack/dlpack_pybind.cc:299:19: error: cannot convert ‘pybind11::none’ to ‘kaldi::DLPackCuSubVector<float>*’ in return
   return py::none();
                   ^
dlpack/dlpack_pybind.cc: In function ‘kaldi::DLPackCuSubMatrix<float>* kaldi::CuSubMatrixFromDLPack(pybind11::capsule*)’:
dlpack/dlpack_pybind.cc:315:19: error: cannot convert ‘pybind11::none’ to ‘kaldi::DLPackCuSubMatrix<float>*’ in return
   return py::none();
                   ^
make: *** [dlpack/dlpack_pybind.o] Error 1
`

It seems it is intended to do this for now

https://github.com/kaldi-asr/kaldi/blob/4807173a3338e90f0c3eee417ac3155875468d6e/src/pybind/dlpack/dlpack_pybind.cc#L289-L301

I'll try to fix it later this week.

The way Kaldi itself works when not compiled for GPU is it expects those
GPU things to actually be on CPU.


On Wed, Jan 8, 2020 at 12:43 PM qindazhu <notifications@github.com> wrote:

> FYI.
>
> g++ -c -std=c++11 -I.. -isystem
> /home/work/kaldi/tools/openfst-1.6.7/include -O1 -Wno-sign-compare -Wall
> -Wno-sign-compare -Wno-unused-local-typedefs -Wno-deprecated-declarations
> -Winit-self -DKALDI_DOUBLEPRECISION=0 -DHAVE_EXECINFO_H=1 -DHAVE_CXXABI_H
> -DHAVE_MKL -I/opt/intel/mkl/include -m64 -msse -msse2 -pthread -g -fPIC
> -I/usr/include/python3.6m -I/usr/local/include/python3.6m
> -I/root/.local/include/python3.6m -O3 -flto -I. -o dlpack/dlpack_pybind.o
> dlpack/dlpack_pybind.cc dlpack/dlpack_pybind.cc: In function
> ‘kaldi::DLPackCuSubVector<float>*
> kaldi::CuSubVectorFromDLPack(pybind11::capsule*)’:
> dlpack/dlpack_pybind.cc:299:19: error: cannot convert ‘pybind11::none’ to
> ‘kaldi::DLPackCuSubVector<float>*’ in return return py::none(); ^
> dlpack/dlpack_pybind.cc: In function ‘kaldi::DLPackCuSubMatrix<float>*
> kaldi::CuSubMatrixFromDLPack(pybind11::capsule*)’:
> dlpack/dlpack_pybind.cc:315:19: error: cannot convert ‘pybind11::none’ to
> ‘kaldi::DLPackCuSubMatrix<float>*’ in return return py::none(); ^ make: ***
> [dlpack/dlpack_pybind.o] Error 1
>
> It seems it is intended to do this for now
>
>
> https://github.com/kaldi-asr/kaldi/blob/4807173a3338e90f0c3eee417ac3155875468d6e/src/pybind/dlpack/dlpack_pybind.cc#L289-L301
>
> I'll try to fix it later this week.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/issues/3825?email_source=notifications&email_token=AAZFLO7QYETCQPBOH6JXARTQ4VKXLA5CNFSM4KECXFFKYY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4IEVC2UA>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLO53ZAARQPXOJEKVEKDQ4VKXLANCNFSM4KECXFFA>
> .
>

Aha, I have done this, but encounter a new error and is trying to fix it now.

`g++ -c -std=c++11 -I.. -isystem /home/work/kaldi/tools/openfst-1.6.7/include -O1 -Wno-sign-compare -Wall -Wno-sign-compare -Wno-unused-local-typedefs -Wno-deprecated-declarations -Winit-self -DKALDI_DOUBLEPRECISION=0 -DHAVE_EXECINFO_H=1 -DHAVE_CXXABI_H -DHAVE_MKL -I/opt/intel/mkl/include -m64 -msse -msse2 -pthread -g -fPIC -I/usr/include/python3.6m -I/usr/local/include/python3.6m -I/root/.local/include/python3.6m -O3 -flto -I. -o util/util_pybind.o util/util_pybind.cc
g++ -std=c++11 -I.. -isystem /home/work/kaldi/tools/openfst-1.6.7/include -O1 -Wno-sign-compare -Wall -Wno-sign-compare -Wno-unused-local-typedefs -Wno-deprecated-declarations -Winit-self -DKALDI_DOUBLEPRECISION=0 -DHAVE_EXECINFO_H=1 -DHAVE_CXXABI_H -DHAVE_MKL -I/opt/intel/mkl/include -m64 -msse -msse2 -pthread -g -fPIC -I/usr/include/python3.6m -I/usr/local/include/python3.6m -I/root/.local/include/python3.6m -O3 -flto -I. -shared -o kaldi_pybind.cpython-36m-x86_64-linux-gnu.so chain/chain_den_graph_pybind.o chain/chain_pybind.o chain/chain_supervision_pybind.o chain/chain_training_pybind.o cudamatrix/cu_device_pybind.o cudamatrix/cu_matrix_pybind.o cudamatrix/cu_vector_pybind.o cudamatrix/cudamatrix_pybind.o dlpack/dlpack_deleter.o dlpack/dlpack_pybind.o dlpack/dlpack_submatrix.o dlpack/dlpack_subvector.o feat/feat_pybind.o feat/feature_pybind.o feat/online_feature_pybind.o feat/wave_reader_pybind.o fst/arc_pybind.o fst/compile_pybind.o fst/fst_pybind.o fst/symbol_table_pybind.o fst/vector_fst_pybind.o fst/weight_pybind.o fstext/kaldi_fst_io_pybind.o kaldi_pybind.o matrix/compressed_matrix_pybind.o matrix/kaldi_matrix_pybind.o matrix/kaldi_vector_pybind.o matrix/matrix_common_pybind.o matrix/matrix_pybind.o matrix/sparse_matrix_pybind.o nnet3/nnet3_pybind.o nnet3/nnet_chain_example_pybind.o nnet3/nnet_common_pybind.o nnet3/nnet_example_pybind.o tests/test_dlpack_subvector.o util/kaldi_io_pybind.o util/table_types_pybind.o util/util_pybind.o -Wl,--no-whole-archive -Wl,-rpath=/home/work/kaldi/src/pybind/../lib  -Wl,-rpath=/home/work/kaldi/tools/openfst-1.6.7/lib -rdynamic -flto -Wl,-rpath=/home/work/kaldi/src/pybind/../../tools/openfst/lib  ../base/libkaldi-base.so  ../chain/libkaldi-chain.so  ../cudamatrix/libkaldi-cudamatrix.so  ../feat/libkaldi-feat.so  ../fstext/libkaldi-fstext.so  ../matrix/libkaldi-matrix.so  ../nnet3/libkaldi-nnet3.so  ../util/libkaldi-util.so ../../tools/openfst/lib/libfstscript.so /home/work/kaldi/tools/openfst-1.6.7/lib/libfst.so -L/opt/intel/mkl/lib/intel64 -Wl,-rpath=/opt/intel/mkl/lib/intel64 -lmkl_intel_lp64  -lmkl_core  -lmkl_sequential    -ldl -lpthread -lm  -lm -lpthread -ldl  ../base/libkaldi-base.so  ../chain/libkaldi-chain.so  ../cudamatrix/libkaldi-cudamatrix.so  ../feat/libkaldi-feat.so  ../fstext/libkaldi-fstext.so  ../matrix/libkaldi-matrix.so  ../nnet3/libkaldi-nnet3.so  ../util/libkaldi-util.so ../../tools/openfst/lib/libfstscript.so
lto1: internal compiler error: in output_die, at dwarf2out.c:8500
Please submit a full bug report,
with preprocessed source if appropriate.
See <http://bugzilla.redhat.com/bugzilla> for instructions.
lto-wrapper: g++ returned 1 exit status
/bin/ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status
make: *** [kaldi_pybind.cpython-36m-x86_64-linux-gnu.so] Error 1
`

--- a/src/pybind/dlpack/dlpack_pybind.cc
+++ b/src/pybind/dlpack/dlpack_pybind.cc
@@ -296,7 +296,7 @@ DLPackCuSubVector<float>* CuSubVectorFromDLPack(py::capsule* capsule) {
                                       tensor->shape[0], managed_tensor);
 #else
   KALDI_ERR << "Kaldi is not compiled with GPU!";
-  return py::none();
+  return nullptr;
 #endif
 }
 
@@ -312,7 +312,7 @@ DLPackCuSubMatrix<float>* CuSubMatrixFromDLPack(py::capsule* capsule) {
                                       tensor->strides[0], managed_tensor);
 #else
   KALDI_ERR << "Kaldi is not compiled with GPU!";
-  return py::none();
+  return nullptr;
 #endif
 }
 

